### ISSUE TITLE
Template rendering error when using bosh cck after failed deploy which introduced a new variable

### ISSUE THREAD

**Issue by rkoster** [cloudfoundry/bosh#2300]
**Describe the bug**
When introducing a new variable via a deploy which fails (on big deployments it can stay in this state for a few days, while waiting for a fix). The resurrector is not able to recreate missing vms. This is making failed deploys extra troublesome operationally because it requires operators to take over the role of the resurrector manually (possible because the `bosh recreate` command does work). 

**To Reproduce**
Run the `./deploy.sh` script from the following repo:
https://github.com/rkoster/reproduce-bosh-template-render-issue

**Expected behavior**
`bosh cck` and resurrector should be able to recreate instances in the described scenario.

**Logs**
```
E, [2021-04-19T14:25:44.476580 #21014] [task:222] ERROR -- DirectorJobRunner: Error resolving problem '14': - Unable to render templates for job 'pre-start-script'. Errors are:
  Expected variable '/bucc/sap-bosh-841/password' to be already versioned in deployment 'sap-bosh-841'
E, [2021-04-19T14:25:44.476634 #21014] [task:222] ERROR -- DirectorJobRunner: /var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh-director-0.0.0/lib/bosh/director/config_server/variables_interpolator.rb:3
6:in `interpolate_template_spec_properties'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/instance_spec.rb:143:in `spec'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/instance_spec.rb:55:in `as_template_spec'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh-director-0.0.0/lib/bosh/director/deployment_plan/steps/render_instance_job_templates_step.rb:17:in `perform'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh-director-0.0.0/lib/bosh/director/step_executor.rb:39:in `block in run_agenda'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh-director-0.0.0/lib/bosh/director/step_executor.rb:36:in `each'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh-director-0.0.0/lib/bosh/director/step_executor.rb:36:in `run_agenda'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh-director-0.0.0/lib/bosh/director/step_executor.rb:24:in `block (4 levels) in run'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh_common-0.0.0/lib/common/thread_formatter.rb:50:in `with_thread_name'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh-director-0.0.0/lib/bosh/director/step_executor.rb:18:in `block (3 levels) in run'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh_common-0.0.0/lib/common/thread_pool.rb:77:in `block (2 levels) in create_thread'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh_common-0.0.0/lib/common/thread_pool.rb:63:in `loop'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/bosh_common-0.0.0/lib/common/thread_pool.rb:63:in `block in create_thread'
/var/vcap/data/packages/director/76c7f4354cac2e6883e2f907827d724c54cc457b/gem_home/ruby/2.6.0/gems/logging-2.2.2/lib/logging/diagnostic_context.rb:474:in `block in create_with_logging_context'
```

**Comment by jpalermo** [comment]
We managed to track down where the problem is originating:

https://github.com/cloudfoundry/bosh/blob/bac774ab8cb921e4dbdd8dc26cb33c875d3ebed4/src/bosh-director/lib/bosh/director/deployment_plan/instance.rb#L149

Here the deployment plan instance is bound to a variable set and the code currently uses the last successful variable set from the deployment.

Previously, this code was using the most recent variable set associated with that particular instance. It was changed as part of a track of work to make start/stop/recreate behavior more consistent (i.e. always go back to the last known good state): https://github.com/cloudfoundry/bosh/commit/53350b010c69a4bc053fe2c7192906e8c3f16723

So in the current code, it's using the last known good variable set, but it's just using the most current state of the VM, which in a failed deploy state, is the new templates rather than the old.

For consistency it would be best if this code path would somehow use the last known good deployment templates, but I don't believe we keep a copy of those, so they'd actually need to be re-rendered possibly. We didn't do much investigation into how we might go about fixing this because it seems like it's going to be rather large change we don't currently have the bandwidth to undertake.

We'd be happy to look at any contributions that try to solve the problem. For now the recommendation is to not leave VMs in a failing state.

Thanks for the comprehensive bug report and reproduction!
